File: src/core/attachmentPreview.js
// Utilities for generating preview thumbnails for chat attachments.
//
// Images → small JPEG data-URL thumbnail (max 320px longest side) so a
//          bubble shows something useful while the full blob loads from
//          IDB / streams over the wire.
// Videos → data-URL of the first frame, captured via <video> + <canvas>.
//
// Kept tiny on purpose. We do not re-encode the full attachment — the
// original blob still gets shipped as-is (with size caps enforced upstream).

export const MAX_THUMB_SIDE = 320;
export const MAX_THUMB_BYTES = 24 * 1024; // ~24 KB JPEG; stays under 32 KB base64

export function classifyFile(file) {
  const mime = String(file?.type || '').toLowerCase();
  if (mime.startsWith('image/')) return 'image';
  if (mime.startsWith('video/')) return 'video';
  if (mime.startsWith('audio/')) return 'audio';
  return 'file';
}

export function formatBytes(n) {
  const num = Number(n) || 0;
  if (num < 1024) return `${num} Б`;
  if (num < 1024 * 1024) return `${(num / 1024).toFixed(1)} КБ`;
  if (num < 1024 * 1024 * 1024) return `${(num / (1024 * 1024)).toFixed(1)} МБ`;
  return `${(num / (1024 * 1024 * 1024)).toFixed(2)} ГБ`;
}

function readImage(blob) {
  return new Promise((resolve, reject) => {
    const url = URL.createObjectURL(blob);
    const img = new Image();
    img.onload = () => {
      resolve({ img, url });
    };
    img.onerror = (err) => {
      try { URL.revokeObjectURL(url); } catch (_) {}
      reject(err);
    };
    img.src = url;
  });
}

function canvasToJpegDataUrl(canvas, quality = 0.72) {
  try {
    return canvas.toDataURL('image/jpeg', quality);
  } catch (_) {
    return null;
  }
}

/**
 * Shrink an image blob into a small JPEG data URL suitable for message
 * preview. Returns `{ thumb, width, height }` of the ORIGINAL image, or
 * null on failure.
 */
export async function buildImageThumbnail(blob) {
  if (!blob || typeof document === 'undefined') return null;
  let loaded;
  try {
    loaded = await readImage(blob);
  } catch (_) {
    return null;
  }
  const { img, url } = loaded;
  try {
    const ow = img.naturalWidth || img.width || 0;
    const oh = img.naturalHeight || img.height || 0;
    if (!ow || !oh) return null;
    const ratio = Math.min(1, MAX_THUMB_SIDE / Math.max(ow, oh));
    const tw = Math.max(1, Math.round(ow * ratio));
    const th = Math.max(1, Math.round(oh * ratio));
    const canvas = document.createElement('canvas');
    canvas.width = tw;
    canvas.height = th;
    const ctx = canvas.getContext('2d');
    if (!ctx) return null;
    ctx.drawImage(img, 0, 0, tw, th);
    let quality = 0.72;
    let thumb = canvasToJpegDataUrl(canvas, quality);
    // Iterate down in quality if the thumb still overshoots.
    while (thumb && thumb.length > MAX_THUMB_BYTES * 1.4 && quality > 0.35) {
      quality -= 0.12;
      thumb = canvasToJpegDataUrl(canvas, quality);
    }
    return { thumb, width: ow, height: oh };
  } finally {
    try { URL.revokeObjectURL(url); } catch (_) {}
  }
}

/**
 * Capture the first frame of a video blob + its duration in seconds.
 * Returns `{ thumb, width, height, duration }` or null.
 */
export async function buildVideoThumbnail(blob) {
  if (!blob || typeof document === 'undefined') return null;
  const url = URL.createObjectURL(blob);
  const video = document.createElement('video');
  video.muted = true;
  video.playsInline = true;
  video.crossOrigin = 'anonymous';
  video.preload = 'metadata';
  video.src = url;
  try {
    await new Promise((resolve, reject) => {
      const onLoad = () => resolve();
      const onErr = () => reject(new Error('video load'));
      video.addEventListener('loadeddata', onLoad, { once: true });
      video.addEventListener('error', onErr, { once: true });
      // Safari sometimes needs an explicit seek to render a frame.
      setTimeout(() => {
        try { video.currentTime = Math.min(0.1, video.duration || 0.1); } catch (_) {}
      }, 30);
    });
    const vw = video.videoWidth || 0;
    const vh = video.videoHeight || 0;
    if (!vw || !vh) return null;
    const ratio = Math.min(1, MAX_THUMB_SIDE / Math.max(vw, vh));
    const tw = Math.max(1, Math.round(vw * ratio));
    const th = Math.max(1, Math.round(vh * ratio));
    const canvas = document.createElement('canvas');
    canvas.width = tw;
    canvas.height = th;
    const ctx = canvas.getContext('2d');
    if (!ctx) return null;
    ctx.drawImage(video, 0, 0, tw, th);
    let quality = 0.72;
    let thumb = canvasToJpegDataUrl(canvas, quality);
    while (thumb && thumb.length > MAX_THUMB_BYTES * 1.4 && quality > 0.35) {
      quality -= 0.12;
      thumb = canvasToJpegDataUrl(canvas, quality);
    }
    return {
      thumb,
      width: vw,
      height: vh,
      duration: Number(video.duration) || 0
    };
  } catch (_) {
    return null;
  } finally {
    try { URL.revokeObjectURL(url); } catch (_) {}
  }
}

export async function buildAttachmentPreview(file) {
  const kind = classifyFile(file);
  if (kind === 'image') {
    const t = await buildImageThumbnail(file);
    return {
      kind,
      thumb: t?.thumb || null,
      width: t?.width || 0,
      height: t?.height || 0,
      duration: 0
    };
  }
  if (kind === 'video') {
    const t = await buildVideoThumbnail(file);
    return {
      kind,
      thumb: t?.thumb || null,
      width: t?.width || 0,
      height: t?.height || 0,
      duration: t?.duration || 0
    };
  }
  return { kind, thumb: null, width: 0, height: 0, duration: 0 };
}

File: lib/core/attachment_preview.dart
// Port of src/core/attachmentPreview.js — thumbnail generation for chat
// attachments.
//
// JS behaviour:
//   image → small JPEG data-URL thumbnail (longest side <= 320px, ~24KB cap)
//   video → data-URL of the first frame (via <video> + <canvas>)
//   other → null
//
// Flutter equivalents require extra packages:
//   - image thumbnails:  package:image for decode + resize + re-encode JPEG.
//   - video thumbnails:  package:video_thumbnail (uses native decoders on
//     Android/iOS, not available on Web/Desktop).
// Neither is in pubspec.yaml yet, so this file ships an API surface compatible
// with the JS version, with stubbed thumbnail generation.
//
// [classifyFile] and [formatBytes] are full ports — they're pure and don't
// need any packages.
//
// TODO(port): add `image` + `video_thumbnail` deps, then fill in the real
// thumbnail pipeline in [buildImageThumbnail] / [buildVideoThumbnail].

import 'dart:typed_data';

const int maxThumbSide = 320;
const int maxThumbBytes = 24 * 1024; // ~24KB JPEG; stays under 32KB base64.

/// Kinds the preview pipeline can produce. Matches the JS string values.
enum AttachmentKind { image, video, audio, file }

String attachmentKindToString(AttachmentKind k) {
  switch (k) {
    case AttachmentKind.image:
      return 'image';
    case AttachmentKind.video:
      return 'video';
    case AttachmentKind.audio:
      return 'audio';
    case AttachmentKind.file:
      return 'file';
  }
}

/// Classify a file into one of four buckets based on its MIME type.
AttachmentKind classifyFile({String? mimeType}) {
  final mime = (mimeType ?? '').toLowerCase();
  if (mime.startsWith('image/')) return AttachmentKind.image;
  if (mime.startsWith('video/')) return AttachmentKind.video;
  if (mime.startsWith('audio/')) return AttachmentKind.audio;
  return AttachmentKind.file;
}

/// Human-readable byte count. Mirrors the JS unit labels (Б/КБ/МБ/ГБ).
String formatBytes(num? n) {
  final num num0 = n ?? 0;
  if (num0 < 1024) return '${num0.toInt()} Б';
  if (num0 < 1024 * 1024) return '${(num0 / 1024).toStringAsFixed(1)} КБ';
  if (num0 < 1024 * 1024 * 1024) {
    return '${(num0 / (1024 * 1024)).toStringAsFixed(1)} МБ';
  }
  return '${(num0 / (1024 * 1024 * 1024)).toStringAsFixed(2)} ГБ';
}

/// Output of the thumbnail builders — a subset of what the React side uses.
class AttachmentPreview {
  const AttachmentPreview({
    required this.kind,
    this.thumb,
    this.width = 0,
    this.height = 0,
    this.duration = 0,
  });

  final AttachmentKind kind;

  /// data:image/jpeg;base64,… URL, or null if thumbnail generation failed /
  /// isn't applicable for this kind.
  final String? thumb;
  final int width;
  final int height;

  /// Seconds, for video.
  final double duration;
}

/// Build an image thumbnail. Returns null on failure or when no decoder is
/// wired up yet.
///
/// TODO(port): decode with package:image, resize longest-side <= 320, then
/// iteratively re-encode JPEG with decreasing quality until size fits
/// [maxThumbBytes] (mirror the 0.72 → 0.60 → 0.48 → 0.36 loop in JS).
Future<AttachmentPreview?> buildImageThumbnail(Uint8List bytes) async {
  if (bytes.isEmpty) return null;
  return null;
}

/// Build a video thumbnail — the first frame as a JPEG. Returns null on
/// failure or when no decoder is wired up yet.
///
/// TODO(port): use package:video_thumbnail to extract frame at t=0.1s, then
/// run through the same JPEG size-iteration loop as [buildImageThumbnail].
Future<AttachmentPreview?> buildVideoThumbnail(Uint8List bytes) async {
  if (bytes.isEmpty) return null;
  return null;
}

/// Dispatch helper — mirror of the JS `buildAttachmentPreview(file)`.
Future<AttachmentPreview> buildAttachmentPreview(
  Uint8List bytes, {
  String? mimeType,
}) async {
  final kind = classifyFile(mimeType: mimeType);
  if (kind == AttachmentKind.image) {
    final t = await buildImageThumbnail(bytes);
    return AttachmentPreview(
      kind: kind,
      thumb: t?.thumb,
      width: t?.width ?? 0,
      height: t?.height ?? 0,
    );
  }
  if (kind == AttachmentKind.video) {
    final t = await buildVideoThumbnail(bytes);
    return AttachmentPreview(
      kind: kind,
      thumb: t?.thumb,
      width: t?.width ?? 0,
      height: t?.height ?? 0,
      duration: t?.duration ?? 0,
    );
  }
  return AttachmentPreview(kind: kind);
}
